drivers/apc_ul_util.c

```diff
@@ -0,0 +1,527 @@
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#include "apc_ul_util.h"
+
+#define APC_UL_PARSE_DEBUG 0
+
+void fletcher16Sum(uint8_t *packet, int length, uint8_t *ret_cs)
+{
+	uint8_t sum1 = 0;
+	uint8_t sum2 = 0;
+	for (int i = 0; i < length; ++i)
+	{
+		sum1 = (sum1 + packet[i]) % 255;
+		sum2 = (sum1 + sum2) % 255;
+	}
+	ret_cs[0] = (uint8_t)sum1;
+	ret_cs[1] = (uint8_t)sum2;
+}
+
+/*  Use ret_cs in the calculation of the checksum instead of starting with 0 */
+void fletcher16SumContinue(uint8_t *packet, int length, uint8_t *ret_cs)
+{
+	uint8_t sum1 = ret_cs[0];
+	uint8_t sum2 = ret_cs[1];
+	for (int i = 0; i < length; ++i)
+	{
+		sum1 = (sum1 + packet[i]) % 255;
+		sum2 = (sum1 + sum2) % 255;
+	}
+	ret_cs[0] = (uint8_t)sum1;
+	ret_cs[1] = (uint8_t)sum2;
+}
+
+/* Compute checksum and store to ret_cs appended to end of packet */
+void calcChecksum(uint8_t *packet, int length, uint8_t *ret_cs)
+{
+	fletcher16Sum(packet, length, ret_cs);
+
+	uint8_t sum1 = ret_cs[0];
+	uint8_t sum2 = ret_cs[1];
+	ret_cs[0] = (uint8_t)(~((sum1 + sum2) % 255) & 255);
+	ret_cs[1] = (uint8_t)(~((ret_cs[0] + sum1) % 255) & 255);
+}
+
+/* Data length is length of the useful part of packet, array size is size of "packet" array */
+int addChecksumToMessage(uint8_t *packet, int data_length, int array_size)
+{
+	if (data_length + 2 > array_size)
+		return 1;
+
+	calcChecksum(packet, data_length, packet + data_length);
+	return 0;
+}
+
+int validateChecksum(uint8_t *packet, int packet_length)
+{
+	uint8_t calc_cs[2];
+	fletcher16Sum(packet, packet_length, calc_cs);
+	if (calc_cs[0] == 0 && calc_cs[1] == 0)
+	{
+		return 0;
+	}
+	return 1;
+}
+
+/* Returns 0 if row is valid */
+int validate_row(microlink_row_t row)
+{
+	if (row.length <= 0)
+	{
+		return 1;
+	}
+	if (validateChecksum(row.array, row.length) != 0)
+	{
+		return 2;
+	}
+	return 0;
+}
+
+/* Converts an FQUN to a string (Either ID 0 is reached or MAX_FQIDS is reached)
+Maximum length of output string is ( 13 * MAX_FQIDS ) + 1 or FQUN_AS_STRING_MAX_LENGTH */
+int fqun_to_string(const fqid_t *fqun, char *dest)
+{
+	int ptr_string = 0;
+	for (int i = 0; i < MAX_FQIDS && fqun[i].id != 0; i++)
+	{
+		fqid_t current_fqid = fqun[i];
+
+		snprintf(dest + ptr_string, 3, "%02X", current_fqid.id);
+		ptr_string += 2;
+
+		switch (current_fqid.type)
+		{
+		case fqid_collection:
+			if (current_fqid.array_length > 0)
+			{
+				snprintf(dest + ptr_string, 12, "[%-3i, %-3i]", current_fqid.array_index % 256, current_fqid.array_length % 256);
```

Comment:
`13` not `12` here, to allow for `\0` always.

Also note the other comment, about adding an argument to check for not overflowing a caller-provided buffer size.